### ISSUE TITLE
Can't resolve z-schema ERROR!!!

### ISSUE THREAD

**Issue by sindhujavs** [makinacorpus/ngx-schema-form#214]
I'm getting the folllowing error:
ERROR in ./node_modules/ngx-schema-form/fesm5/ngx-schema-form.js
Module not found: Error: Can't resolve 'z-schema' in '..../node_modules/ngx-schema-form/fesm5'

Could anyone help. Thanks in advance.

**Comment by WhileTrueEndWhile** [comment]
Maybe that sounds too easy now, but did you install Z-Schema?

_package.json_
```
...
    "dependencies": {
        ...
         "z-schema": "^3.17.0"
...
```

I don't know if it will be installed, because the schema validator is interchangeable.

**Comment by sindhujavs** [comment]
Yup i did but didnt work.

**Comment by WhileTrueEndWhile** [comment]
Does it work with the old "angular2" version?

**Comment by novitskiy-aleksei** [comment]
I'm using latest ngx-schema-form with angular6 + SystemJs and having similar issue

![image](https://user-images.githubusercontent.com/2851476/41910443-565cb5d2-7952-11e8-95d8-b9ce6de83303.png)

Have any thoughts about this?

**Comment by shohrukh92** [comment]
I've these versions:
`"ngx-schema-form": "^2.1.0-beta.1"`
`"z-schema": "^3.22.0"`
`"@angular/*": "^6.0.3"`
and it works without any errors

**Comment by gpessa** [comment]
same here, If I fix it I get other deps missing

**Comment by htmoia** [comment]
Same with 
```json
"ngx-schema-form": "2.2.0-beta.1",
"@angular/core": "5.2.11",
"@ionic-native/core": "4.10.0",
"ionic-angular": "3.9.2"
```

`Error: Cannot find module "z-schema" ngx-schema-form.js:1 `

**Comment by ebrehault** [comment]
Just make sure you have the same deps as here: https://github.com/guillotinaweb/ngx-schema-form/blob/master/package.json#L15 and it will run fine (that's how Travis is doing its install at the moment and the build is successful).

**Comment by tqduy** [comment]
Hi @ebrehault ,
I have an issue when upgrading from angular2-schema-form to ngx-schema-form where I add the z-schema package
I have a schema json:
`"component": {
                "type": "string",
                "widget": "select",
                "description": "Component",
                "oneOf": [
                  {
                    "description": "Audible Warning Devices",
                    "enum": [
                      "AudibleWarningDevices"
                    ]
                  },
                  {
                    "description": "Bonnet Release Cable",
                    "enum": [
                      "BonnetReleaseCable"
                    ]
                  }
                ]
              }`

I face an error when my widget try to get the initial data from this.FormProperty.value
![image](https://user-images.githubusercontent.com/11729162/72314942-794f8a80-36c3-11ea-8925-b54a2065594a.png)

Do you have any idea about this? I think it is about the parsing approach is changed because of z-schema

**Comment by ebrehault** [comment]
Sorry, I never saw this problem before.
Maybe you could try to use the source version of ngx-schema-form so you can debug and see what happens.

**Comment by tqduy** [comment]
> Sorry, I never saw this problem before.
> Maybe you could try to use the source version of ngx-schema-form so you can debug and see what happens.
thank you for response.
Can you give me some clue so that I can track down about how the ControlWidget.formProperty is binding while rendering between the steps the <sf-
form> to call widget class? 
![image](https://user-images.githubusercontent.com/11729162/72335375-78355200-36f1-11ea-97e2-6da5e06a71a4.png)

The above error lead the formProperty._value = "" and cannot render the existing value loaded from API server

**Comment by ebrehault** [comment]
I cannot provide any support at the moment as I am pretty busy with other matters.

@daniele-pecora, would you have time to have a quick look?

**Comment by aman-di-fu** [comment]
@tqduy 
I had this once when the `oneOf` didn't contain an empty value selection.
Simply add something like this to your `oneOf` schema :
```json
        {
          "description": "-- Make your selection --",
          "enum": [""]
        }
```
For me this behavior seems logic because the field's value has to be one of described by `oneOf` but initial it is `null` or `` (empty)

**Comment by tqduy** [comment]
> @tqduy
> I had this once when the `oneOf` didn't contain an empty value selection.
> Simply add something like this to your `oneOf` schema :
> 
> ```json
>         {
>           "description": "-- Make your selection --",
>           "enum": [""]
>         }
> ```
> 
> For me this behavior seems logic because the field's value has to be one of described by `oneOf` but initial it is `null` or `` (empty)

Thank you for response.
Your solution can solve the error message. But I still cannot received the initial data in the widget to render
I will explain more: I have a component named A.component.html which used sf-form which schema and schemaModel come from API server

`
<sf-form *ngIf="showSchema" [schema]="schema" (onChange)="updateSchemaValue($event.value)"
        [validators]="fieldValidators" [actions]="actions" [(model)]="schemaModel">

`

the schema is an array schema widget:
 ` export class IonicArrayWidgetComponent extends ArrayWidget implements AfterViewInit { }`
with HTML : `   <sf-form-element [attr.id]="i" [formProperty]="itemProperty"></sf-form-element> `

==> each element from the array widget is the group of ion-selectable : ` export class IonicSelectWithSearchComponent extends SelectWidget { } `

==> The issue is about I cannot render the data for each IonicSelectWithSearchComponent because the value is always "". I go back to the IonicArrayWidgetComponent and see that the formProperty._value is "" as well. So I think there must be some mistake at the component A which sf-form cannot passing the schemaModel (aka schema data) to the widget

**Comment by gabriellupu** [comment]
```
npm i jexl --save
npm i z-schema --save
```
fixed it for me. using `"ngx-schema-form": "^2.4.3",`

**Comment by daniele-pecora** [comment]
> I cannot provide any support at the moment as I am pretty busy with other matters.
> 
> @daniele-pecora, would you have time to have a quick look?

I'll have a look asap

**Comment by daniele-pecora** [comment]
> > @tqduy
> > I had this once when the `oneOf` didn't contain an empty value selection.
> > Simply add something like this to your `oneOf` schema :
> > ```json
> >         {
> >           "description": "-- Make your selection --",
> >           "enum": [""]
> >         }
> > ```
> > 
> > 
> > For me this behavior seems logic because the field's value has to be one of described by `oneOf` but initial it is `null` or `` (empty)
> 
> @aman-di-fu
> Thank you for response.
> Your solution can solve the error message. But I still cannot received the initial data in the widget to render
> I will explain more: I have a component named A.component.html which used sf-form which schema and schemaModel come from API server
> 
> `<sf-form *ngIf="showSchema" [schema]="schema" (onChange)="updateSchemaValue($event.value)"`
> ` [validators]="fieldValidators" [actions]="actions" [(model)]="schemaModel">`
> 
> the schema is an array schema widget:
> ` export class IonicArrayWidgetComponent extends ArrayWidget implements AfterViewInit { }`
> with HTML : ` <sf-form-element [attr.id]="i" [formProperty]="itemProperty"></sf-form-element>`
> 
> ==> each element from the array widget is the group of ion-selectable : `export class IonicSelectWithSearchComponent extends SelectWidget { }`
> 
> ==> The issue is about I cannot render the data for each IonicSelectWithSearchComponent because the value is always "". I go back to the IonicArrayWidgetComponent and see that the formProperty._value is "" as well. So I think there must be some mistake at the component A which sf-form cannot passing the schemaModel (aka schema data) to the widget

I'd recommend using :
```
               [model]="model"
               (onChangeModel)="model=$event"
```
instead of 
```

               [(model)]="model"
```

this solves the issue that the array model is not pre-filled.

@ebrehault I think this is the problem...

**Comment by ebrehault** [comment]
Agreed.
I am even wondering if we should add a warning message (in dev mode only) when people use `changeModel` (the implicit output used with `[(model)]` because it is causing so many confusion.

**Comment by daniele-pecora** [comment]
@ebrehault We already struggled on this with #233

**Comment by ebrehault** [comment]
yes, true.
So if we want to restrict two-way binding to `ngModel` only, we just have to remove  the `modelChange` output.

**Comment by daniele-pecora** [comment]
@ebrehault This would be a breaking change?

**Comment by ebrehault** [comment]
Yes, and we will mention it in the changelog. I think that's fine, it will avoid confusion in the future.
